### PR TITLE
Some QoL improvements for diag tests

### LINKED ISSUE #1202: Build succeeds even when generator fails
**Describe the bug**
The build should fail when the generator encounters an error

**To Reproduce**
```csharp
namespace CSWinRT_AuthoringPlayground
{
    public enum MyEnum
    {
        A=1,
        B
    }

    public struct MyStruct
    {
        public Enum x;
    }
}
```

Build output:
image

**Expected behavior**
Build fails

**Version Info**
CsWinRT 1.6.4

### PR DESCRIPTION
**Pull request by asklar**
- Fixes DiagnosticTests! There were a few non-passing tests so I debugged into it and saw that we're failing because in some codepaths we didn't have an assembly name or version. Passing these requires a context options object.
- DiagnosticTests can now pass a context with options, so we can remove special casing we had for cswinrt generator when running inside a test host
- When tests failed, the inner exception was not being captured, so the error just said "...But  found:" and no additional information about the exception in the generator was provided
- Setting ExitCode upon error, which should cause builds to fail when the generator fails to produce. Fixes #1202 
- Fixes some typos

image

### CODE REVIEW

File: src/Tests/DiagnosticTests/Helpers.cs [hunk 1]
```diff
@@ -14,25 +15,25 @@ public sealed partial class UnitTesting
...
         /// <returns></returns>
-        private static GeneratorDriver CreateDriver(Compilation compilation, params ISourceGenerator[] generators)
+        private static GeneratorDriver CreateDriver(Compilation compilation, AnalyzerConfigOptionsProvider? options, params ISourceGenerator[] generators)
             => CSharpGeneratorDriver.Create(
                 generators: ImmutableArray.Create(generators),
                 additionalTexts: ImmutableArray<AdditionalText>.Empty,
                 parseOptions: (CSharpParseOptions)compilation.SyntaxTrees.First().Options,
-                optionsProvider: null); 
+                optionsProvider: options); 
         // todo: pass the CsWinRTComponent config option here so we don't have to comment out the check in the source generator
```
Comment (manodasanW) [inline src/Tests/DiagnosticTests/Helpers.cs:37]: I think you can drop this comment as your change addresses it.

**Review by manodasanW — COMMENTED** [review]
Thanks @asklar for these changes.

File: src/Tests/DiagnosticTests/Helpers.cs (hunk 1 above)
Comment (j0shuams) [reply src/Tests/DiagnosticTests/Helpers.cs:37]: +1

**Review by j0shuams — APPROVED** [review]
Looks good to me, just needs some feedback on the comments

File: src/Tests/DiagnosticTests/UnitTesting.cs [hunk 2]
```diff
@@ -19,17 +59,13 @@ public sealed partial class UnitTesting
...
             if (WinRTDiagnostics.Any())
             {
-                string foundDiagnostics = "";
-                foreach (var d in WinRTDiagnostics)
-                {
-                    foundDiagnostics += d.Descriptor.Description + "\n";
-                }
-                throw new System.Exception("Expected no diagnostics. But found:" + foundDiagnostics);
+                var foundDiagnostics = string.Join("\n", WinRTDiagnostics.Select(x => x.GetMessage()));
+                throw new SuccessException("Expected no diagnostics. But found:\n" + foundDiagnostics);
```
[the remark this answers is not in the document]
Comment (asklar) [reply src/Tests/DiagnosticTests/UnitTesting.cs:68]: SuccessException is "Thrown when an assertion failed."; it's the NUnit exception type for when something fails (the name is confusing :) )

**Review by j0shuams — APPROVED** [review]

**Pull request merged**